01_TS/5-power-user-type.ts

```diff
@@ -0,0 +1,79 @@
+/* eslint-disable @typescript-eslint/no-inferrable-types */
+/* Exercise:
+    Define type "PowerUser" which should have all fields
+    from both "User" and "Admin" (except for "type" property),
+    and also have "type" property equals to "powerUser" without duplicating
+    all the fields in the code.
+    Hint: use utility types.
+*/
+
+interface User {
+    type: 'user';
+    name: string;
+    age: number;
+    occupation: string;
+}
+
+interface Admin {
+    type: 'admin';
+    name: string;
+    age: number;
+    role: string;
+}
+
+type PowerUser = Omit<User, 'type'> & Omit<Admin, 'type'> & {type: 'powerUser'}; /* [Declare type "PowerUser" here.] */
+
+export type Person = User | Admin | PowerUser;
+
+export const persons: Person[] = [
+    { type: 'user', name: 'Max Mustermann', age: 25, occupation: 'Chimney sweep' },
+    { type: 'admin', name: 'Jane Doe', age: 32, role: 'Administrator' },
+    { type: 'user', name: 'Kate Müller', age: 23, occupation: 'Astronaut' },
+    { type: 'admin', name: 'Bruce Willis', age: 64, role: 'World saver' },
+    {
+        type: 'powerUser',
+        name: 'Nikki Stone',
+        age: 45,
+        role: 'Moderator',
+        occupation: 'Cat groomer'
+    }
+];
+
+function isAdmin(person: Person): person is Admin {
+    return person.type === 'admin';
+}
+
+function isUser(person: Person): person is User {
+    return person.type === 'user';
+}
+
+function isPowerUser(person: Person): person is PowerUser {
+    return person.type === 'powerUser';
+}
+
+export function logPerson(person: Person) {
+    let additionalInformation: string = '';
+    if (isAdmin(person)) {
+        additionalInformation = person.role;
+    }
+    if (isUser(person)) {
+        additionalInformation = person.occupation;
+    }
+    if (isPowerUser(person)) {
+        additionalInformation = `${person.role}, ${person.occupation}`;
+    }
+    console.log(`${person.name}, ${person.age}, ${additionalInformation}`);
+}
+
+console.log('Admins:');
+persons.filter(isAdmin).forEach(logPerson);
+
+console.log();
+
+console.log('Users:');
+persons.filter(isUser).forEach(logPerson);
+
+console.log();
+
+console.log('Power users:');
+persons.filter(isPowerUser).forEach(logPerson);
```

Comment:
Please add tests to your code (attached with the task itself).

Reply:
Added tests folder

Reply:
Added tetst folder